enzyme/test/TypeAnalysis/smaxsmin.ll

```diff
@@ -0,0 +1,29 @@
+; RUN: if [ %llvmver -lt 16 ]; then %opt < %s %loadEnzyme -print-type-analysis -type-analysis-func=smax -o /dev/null | FileCheck %s; fi
+; RUN: %opt < %s %newLoadEnzyme -passes="print-type-analysis" -type-analysis-func=smaxsmin -S -o /dev/null | FileCheck %s
+
+declare void @_mlir_memref_to_llvm_free(ptr)
+
+declare ptr @_mlir_memref_to_llvm_alloc(i64)
+
+declare i64 @llvm.smin.i64(i64, i64)
+
+declare i32 @llvm.smax.i32(i32, i32)
+
+define internal void @smaxsmin(ptr %0) {
+entry:
+  %1 = tail call ptr @_mlir_memref_to_llvm_alloc(i64 72)
+  %2 = ptrtoint ptr %1 to i64
+  %3 = add i64 %2, 63
+  %4 = and i64 %3, -64
+  %5 = inttoptr i64 %4 to ptr
+  %6 = load i32, ptr %5, align 64
+  %7 = tail call i32 @llvm.smax.i32(i32 %6, i32 0)
+  %8 = zext nneg i32 %7 to i64
+  %9 = tail call i64 @llvm.smin.i64(i64 %8, i64 3)
+  %10 = getelementptr double, ptr %0, i64 %9
+  %11 = load double, ptr %10, align 8
+  ret void
+}
+
+; CHECK-NOT: %7 = tail call i32 @llvm.smax.i32(i32 %6, i32 0): {}
```

Comment:
@wsmoses Here is the test not passing, let me know if you see anything that could help.

Reply:
Yeah from your snippet here I actually agree with Type analysis that there isn't enough info to ascertain the types.

There are several potential ways to resolve this, but it would probably be easiest to see the MLIR which is lowered to figure out which.

For example presently the arg ptr %0 has no information about what it is [eg whether it is a pointer, int etc].

Since there is a gep %0, %9, that may be one good way to inject information. If it was a gep inbounds %0, %9 we can deduce that %9 must be an integer [and then with this PR it propagates upward perfectly]. However without the inbounds, it is possible that %0 is actually null and the pointer is %9 [which can come from an inttoptr]. This latter case actually is emitted by LLVM in practice, in particular for virtual C++ functions.

Another route may be attaching type information to _mlir_memref_to_llvm_alloca to say it is a memref of int, so all loads / stores are ints. Alternatively one can put TBAA on the load/store generated by the memref.

etc.

I'd say for now to move this test out of the PR into an issue, so we can get this PR merged, and then follow up for this next.

Reply:
FYI, the results with GEP being marked inbounds:

```
wmoses@beast:~/git/Enzyme/enzyme/build15 (add_max_min_type_analysis) $  /home/wmoses/llvms/llvm15/buildD/./bin/opt < /home/wmoses/git/Enzyme/enzyme/test/TypeAnalysis/smaxsmin.ll  -load-pass-plugin=/home/wmoses/git/Enzyme/enzyme/build15/Enzyme/LLVMEnzyme-15.so -load=/home/wmoses/git/Enzyme/enzyme/build15/Enzyme/LLVMEnzyme-15.so --enzyme-attributor=0 -passes="print-type-analysis" -type-analysis-func=smaxsmin -S -o /dev/null
smaxsmin - {} |{[-1]:Pointer}:{} 
ptr %0: {[-1]:Pointer}
entry
  %1 = tail call ptr @_mlir_memref_to_llvm_alloc(i64 72): {}
  %2 = ptrtoint ptr %1 to i64: {}
  %3 = add i64 %2, 63: {}
  %4 = and i64 %3, -64: {[-1]:Pointer, [-1,0]:Integer, [-1,1]:Integer, [-1,2]:Integer, [-1,3]:Integer}
  %5 = inttoptr i64 %4 to ptr: {[-1]:Pointer, [-1,0]:Integer, [-1,1]:Integer, [-1,2]:Integer, [-1,3]:Integer}
  %6 = load i32, ptr %5, align 64: {[-1]:Integer}
  %7 = tail call i32 @llvm.smax.i32(i32 %6, i32 0): {[-1]:Integer}
  %8 = zext i32 %7 to i64: {[-1]:Integer}
  %9 = tail call i64 @llvm.smin.i64(i64 %8, i64 3): {[-1]:Integer}
  %10 = getelementptr inbounds double, ptr %0, i64 %9: {[-1]:Pointer}
  %11 = load double, ptr %10, align 8: {}
  ret void: {}
  ```

Reply:
Thanks a lot @wsmoses for all the explanation!